model_zoo/CorrectDNN/src/Correct.py

```diff
@@ -0,0 +1,209 @@
+                                      num_negs=num_negs,
+                                      embedding_initializer=embedding_initializer,
+                                      **kwargs)
+        self.similarity_score = similarity_score
+        self.embedding_dim = embedding_dim
+        self.num_items = num_items + 1
+        self.user_id_field = user_id_field
+        self.user_history_field = user_history_field
+        self.embedding_layer = EmbeddingDictLayer(feature_map, embedding_dim)
+        self.behavior_aggregation = BehaviorAggregator(embedding_dim, 
+                                                       gamma=gamma,
+                                                       aggregator=aggregator, 
+                                                       dropout_rate=attention_dropout)
+        self.enable_bias = enable_bias
+        if self.enable_bias:
+            self.user_bias = EmbeddingLayer(feature_map, 1,
+                                            disable_sharing_pretrain=True, 
+                                            required_feature_coluCorrect=[user_id_field])
+            self.item_bias = EmbeddingLayer(feature_map, 1, 
+                                            disable_sharing_pretrain=True, 
+                                            required_feature_coluCorrect=[item_id_field])
+            self.global_bias = nn.Parameter(torch.zeros(1))
+        self.dropout = nn.Dropout(net_dropout)
+        self.arr_A = torch.zeros(self.num_items).to(self.device)
+        self.arr_B = torch.ones(self.num_items).to(self.device) * 100
+        self.step = 0
+        self.lr = learning_rate
+        self.compile(lr=learning_rate, **kwargs)
+
+
+    def sample_in_batch(self, samples, batch):
+        sampling_probs = self.arr_B  # uniform sampling
+        sampled_items = []
+        # print(samples[:, 0].max().numpy(), self.num_items)
+        pos_items = samples[:, 0].numpy()
+        probs = np.array(sampling_probs)
+        probs[pos_items] = 0
+        probs[-1] = 0
+        probs = probs / np.sum(probs)  # renomalize to sum 1
+        neg_in_unis = set(pos_items)
+        # Reconstruct the negative samples
+        for i in range(batch):
+            pos_item = samples[i][0]
+            neg_in_batch = np.delete(pos_items, np.where(pos_items==pos_item.numpy()))
+            neg_in_uni = np.random.choice(self.num_items, size=self.num_negs-len(neg_in_batch), replace=True, p=probs)
+            negs = np.concatenate([np.array([pos_item]), neg_in_batch, neg_in_uni]).tolist()
+            sampled_items += negs
+            for neg in neg_in_uni:
+                neg_in_unis.add(neg)
+
+        # Update array A and B
+        self.step += 1
+        for item in neg_in_unis:
+            self.arr_B[item] = (1-self.lr)*self.arr_B[item] + self.lr*(self.step-self.arr_A[item])
+            self.arr_A[item] = self.step
+
+        return torch.tensor(sampled_items).to(samples.device)
+
+    def forward(self, inputs):
```

Comment:
@Tokkiu 
My guess is that if we only use in-batch negatives, we don't need to do any sampling. If we implement the forward statement as follows, we can implement a sampled softmax loss using in-batch negatives. (please refer `else` clause.)

```
    def forward(self, inputs):
        """
        Inputs: [user_dict, item_dict, label]
        """
        user_dict, item_dict, labels = inputs[0:3]
        user_vecs = self.user_tower(user_dict)
        # print('forward user_vecs', user_vecs)
        user_vecs = self.user_dropout(user_vecs)
        item_vecs = self.item_tower(item_dict)
        # print('forward item_vecs', item_vecs)
        item_vecs = self.item_dropout(item_vecs)
        if not self.in_batch_negatives:
            y_pred = torch.bmm(
                item_vecs.view(-1, self.num_negs + 1, self.output_dim),
                user_vecs.unsqueeze(-1),
            ).squeeze(-1)
            loss = self.get_total_loss(y_pred, labels)
            return_dict = {"loss": loss, "y_pred": y_pred}
        else:
            # item_vecs: [B x (1+num_negs) x output_dim]
            # user_vecs: [B x output_dim]
            batch_items = item_dict["item_id"].view(
                user_vecs.size(0), self.num_negs + 1
            )
            pos_items = batch_items[:, 0].cpu().numpy()
            in_batch_freqs = Counter(pos_items) # Dict of {item_id: freq} # len(in_bathc_freqs) = batch_size
            logQ_correction = (
                torch.log(
                    torch.tensor(
                        [in_batch_freqs[i] for i in pos_items], dtype=torch.float32
                    )
                    + 1e-6
                )
                .reshape((1, -1))
                .to(self.device)
            )
            viewed_item_vecs = item_vecs.view(-1, self.num_negs + 1, self.output_dim)
            pos_item_vecs = viewed_item_vecs[:, 0, :] # positive item만 slicing
            score = self.cos(user_vecs.unsqueeze(1), pos_item_vecs.unsqueeze(0)) # B x B
            score -= logQ_correction
            labels = torch.arange(score.size(0)).long().to(self.device)

            loss = self.loss_fn(score, labels)  # need to add regularizaiton loss
            return_dict = {"loss": loss}

        return return_dict
```

I'll share the model_config.yaml file and full src code on slack for reproduction.